consensus/polybft/polybft.go

```diff
@@ -0,0 +1,662 @@
+		return fmt.Errorf("failed to validate header for block %d. error = %w", blockNumber, err)
+	}
+
+	validators, err := p.GetValidators(blockNumber-1, parents)
+	if err != nil {
+		return fmt.Errorf("failed to validate header for block %d. could not retrieve block validators:%w", blockNumber, err)
+	}
+
+	// decode the extra field and validate the signatures
+	extra, err := GetIbftExtra(header.ExtraData)
+	if err != nil {
+		return fmt.Errorf("failed to verify header for block %d. get extra error = %w", blockNumber, err)
+	}
+
+	if extra.Committed == nil {
+		return fmt.Errorf(
+			"failed to verify signatures for block %d because signatures are nil. Block hash: %v",
+			blockNumber,
+			header.Hash,
+		)
+	}
+
+	if err := extra.Committed.VerifyCommittedFields(validators, header.Hash); err != nil {
+		return fmt.Errorf("failed to verify signatures for block %d. Block hash: %v", blockNumber, header.Hash)
+	}
+
+	// validate the signatures for parent (skip block 1 because genesis does not have committed)
+	if blockNumber > 1 {
+		if extra.Parent == nil {
+			return fmt.Errorf(
+				"failed to verify signatures for parent of block %d because signatures are nil. Parent hash: %v",
+				blockNumber,
+				parent.Hash,
+			)
+		}
+
+		parentValidators, err := p.GetValidators(blockNumber-2, parents)
+		if err != nil {
+			return fmt.Errorf(
+				"failed to validate header for block %d. could not retrieve parent validators:%w",
+				blockNumber,
+				err,
+			)
+		}
+
+		if err := extra.Parent.VerifyCommittedFields(parentValidators, parent.Hash); err != nil {
+			return fmt.Errorf("failed to verify signatures for parent of block %d. Parent hash: %v", blockNumber, parent.Hash)
+		}
+	}
+
+	return nil
+}
+
+func (p *Polybft) CheckIfStuck(num uint64) (uint64, bool) {
+	if !p.isSynced() {
+		// we are currently syncing new data, for sure we are stuck.
+		// We can return 0 here at least for now since that value is only used
+		// for the open telemetry tracing.
+		return 0, true
+	}
+
+	// Now, we have to check if the current value of the round 'num' is lower
+	// than our currently synced block.
+	currentHeader := p.blockchain.CurrentHeader().Number
+	if currentHeader > num {
+		// at this point, it will exit the sync process and start the fsm round again
+		// (or sync a small number of blocks) to start from the correct position.
+		return currentHeader, true
+	}
+
+	return 0, false
+}
+
+func (p *Polybft) GetValidators(blockNumber uint64, parents []*types.Header) (AccountSet, error) {
+	return p.validatorsCache.GetSnapshot(blockNumber, parents)
+}
+
+// ProcessHeaders updates the snapshot based on the verified headers
+func (p *Polybft) ProcessHeaders(_ []*types.Header) error {
+	// Not required
+	return nil
+}
+
+// GetBlockCreator retrieves the block creator (or signer) given the block header
+func (p *Polybft) GetBlockCreator(_ *types.Header) (types.Address, error) {
+	return types.Address(p.key.Address()), nil
+}
```

Comment:
This method is not doing what it's supposed to be doing.

It should extract the proposer from the extra data field, and not return the current node's address

Reply:
Fixed 604c2ffe5a7a4ea063fb3eabd56c9cae14ca056a